### PR TITLE
Search Messages

### PR DESCRIPTION
**Pull request by MatkoTom**
# Description

Please include a summary of the change and which issue is fixed. Please also include relevant
motivation and context. List any dependencies that are required for this change.

Implementation of the search messages functionality. Users can now search for all chat messages on the chat room screen. By clicking on one of them, the user should be taken to the room and navigated to the message.

## Type of change

Please delete options that are not relevant.

- [x] New feature (non-breaking change which adds functionality)

# How Has This Been Tested?

Please describe the tests that you ran to verify your changes. Provide instructions so we can
reproduce. Please also list any relevant details for your test configuration. All testing was done
manually by following, plainly, functionality requirements. If there are some special edge cases or
special ways that things need to be tested they will be mentioned below.

Dev testing

**Test Configuration**:

* Firmware version: G980FXXSFFVHA
* Hardware: SAMSUNG Galaxy S20
* SDK: Android 13

# Checklist:

- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation
- [x] My changes generate no new warnings
- [x] Any dependent changes have been merged and published in downstream modules
JPEG_20230801_131513_3556120417891599100
JPEG_20230801_131513_8246004820184347840
5b7c5872)

### CODE REVIEW

File: app/src/main/java/com/clover/studio/spikamessenger/ui/main/chat/ChatMessagesFragment.kt [hunk 1]
```diff
@@ -227,32 +231,53 @@ class ChatMessagesFragment : BaseFragment(), ChatOnBackPressed {
...
+                    setUpAdapter()
+
+                    // Clear notifications for this room
+                    NotificationManagerCompat.from(requireContext())
+                        .cancel(roomWithUsers.room.roomId)
+                }
+
+                Resource.Status.ERROR -> Toast.makeText(
+                    activity,
+                    "Failed to load room data",
```
Comment (ivolarevic) [inline app/src/main/java/com/clover/studio/spikamessenger/ui/main/chat/ChatMessagesFragment.kt:274]: Maybe extract this value?

File: app/src/main/java/com/clover/studio/spikamessenger/ui/main/chat/ChatMessagesFragment.kt [hunk 2]
```diff
@@ -227,32 +231,53 @@ class ChatMessagesFragment : BaseFragment(), ChatOnBackPressed {
...
+                    Toast.LENGTH_SHORT
+                ).show()
+
+                else -> Toast.makeText(activity, "Other error", Toast.LENGTH_SHORT).show()
```
Comment (ivolarevic) [inline app/src/main/java/com/clover/studio/spikamessenger/ui/main/chat/ChatMessagesFragment.kt:278]: Same as above

File: app/src/main/java/com/clover/studio/spikamessenger/ui/main/chat/ChatMessagesFragment.kt [hunk 3]
```diff
@@ -227,32 +231,53 @@ class ChatMessagesFragment : BaseFragment(), ChatOnBackPressed {
...
+                    roomWithUsers = it.responseData?.roomWithUsers!!
 
-        // Clear notifications for this room
-        NotificationManagerCompat.from(requireContext()).cancel(roomWithUsers.room.roomId)
+                    if (Const.JsonFields.PRIVATE == roomWithUsers.room.type) {
+                        user =
+                            roomWithUsers.users.firstOrNull { user -> user.id.toString() != localUserId.toString() }
+                    }
+
+                    // Timber.d("Load check: ChatMessagesFragment view created")
```
Comment (ivolarevic) [inline app/src/main/java/com/clover/studio/spikamessenger/ui/main/chat/ChatMessagesFragment.kt:253]: Maybe we can remove this comment if we don't use it anymore for testing

File: app/src/main/java/com/clover/studio/spikamessenger/ui/main/chat/ChatMessagesFragment.kt [hunk 4]
```diff
@@ -622,6 +647,24 @@ class ChatMessagesFragment : BaseFragment(), ChatOnBackPressed {
                             bindingSetup.rvChat.layoutManager?.onRestoreInstanceState(listState)
                             shouldScroll = false
                         }
+
+                        Timber.d("Message search id value: $messageSearchId")
```
Comment (ivolarevic) [inline app/src/main/java/com/clover/studio/spikamessenger/ui/main/chat/ChatMessagesFragment.kt:651]: Maybe you can comment out this?

File: app/src/main/java/com/clover/studio/spikamessenger/ui/main/chat/ChatMessagesFragment.kt [hunk 5]
```diff
@@ -622,6 +647,24 @@ class ChatMessagesFragment : BaseFragment(), ChatOnBackPressed {
...
+                        if (messageSearchId != 0) {
+                            if (messagesRecords.firstOrNull { messageAndRecords -> messageAndRecords.message.id == messageSearchId } != null) {
+                                Timber.d("Message found message search id: $messageSearchId")
```
Comment (ivolarevic) [inline app/src/main/java/com/clover/studio/spikamessenger/ui/main/chat/ChatMessagesFragment.kt:654]: Same as above

File: app/src/main/java/com/clover/studio/spikamessenger/ui/main/chat/ChatMessagesFragment.kt [hunk 6]
```diff
@@ -622,6 +647,24 @@ class ChatMessagesFragment : BaseFragment(), ChatOnBackPressed {
...
+                                val position =
+                                    messagesRecords.indexOfFirst { messageAndRecords -> messageAndRecords.message.id == messageSearchId }
+                                Timber.d("Message found position: $position")
```
Comment (ivolarevic) [inline app/src/main/java/com/clover/studio/spikamessenger/ui/main/chat/ChatMessagesFragment.kt:657]: Same as above

File: app/src/main/java/com/clover/studio/spikamessenger/ui/main/chat/ChatMessagesFragment.kt [hunk 7]
```diff
@@ -622,6 +647,24 @@ class ChatMessagesFragment : BaseFragment(), ChatOnBackPressed {
...
+                                scrollToPosition = position
+                                if (position != -1) {
+                                    bindingSetup.rvChat.smoothScrollToPosition(position)
+                                }
+                                messageSearchId = 0
+                            } else {
+                                Timber.d("Message not found, fetch next set")
```
Comment (ivolarevic) [inline app/src/main/java/com/clover/studio/spikamessenger/ui/main/chat/ChatMessagesFragment.kt:664]: Same as above

File: app/src/main/java/com/clover/studio/spikamessenger/ui/main/chat/ChatScreenActivity.kt [hunk 8]
```diff
@@ -82,19 +84,19 @@ class ChatScreenActivity : BaseActivity() {
...
-                    it.responseData?.roomWithUsers?.let { roomWithUsers ->
-                        replaceChatScreenActivity(this, roomWithUsers)
-                    }
-                }
-
-                Resource.Status.ERROR -> Timber.d("Failed to fetch room data")
-                else -> Timber.d("Other error")
-            }
-        })
+//        viewModel.roomDataListener.observe(this, EventObserver {
```
Comment (ivolarevic) [inline app/src/main/java/com/clover/studio/spikamessenger/ui/main/chat/ChatScreenActivity.kt:87]: Can we remove this or? Why is this removed?

File: app/src/main/java/com/clover/studio/spikamessenger/ui/main/contact_details/ContactDetailsFragment.kt [hunk 9]
```diff
@@ -101,7 +101,9 @@ class ContactDetailsFragment : BaseFragment() {
             when (it.status) {
                 Resource.Status.SUCCESS -> {
                     Timber.d("Room with users = ${it.responseData}")
-                    activity?.let { parent -> startChatScreenActivity(parent, it.responseData!!) }
+                    activity?.let { parent -> startChatScreenActivity(parent,
+                        it.responseData?.room?.roomId!!
```
Comment (ivolarevic) [inline app/src/main/java/com/clover/studio/spikamessenger/ui/main/contact_details/ContactDetailsFragment.kt:105]: What if roomId is null?

File: app/src/main/java/com/clover/studio/spikamessenger/ui/main/rooms/RoomsFragment.kt [hunk 10]
```diff
@@ -47,42 +54,79 @@ class RoomsFragment : BaseFragment() {
         return binding.root
     }
 
+    private fun initializeViews() = with(binding) {
```
Comment (ivolarevic) [inline app/src/main/java/com/clover/studio/spikamessenger/ui/main/rooms/RoomsFragment.kt:57]: Good work Matko!! ✅✅

File: app/src/main/java/com/clover/studio/spikamessenger/ui/main/rooms/RoomsFragment.kt [hunk 11]
```diff
@@ -47,42 +54,79 @@ class RoomsFragment : BaseFragment() {
...
+
+        btnSearchMessages.setOnClickListener {
+            rvMessages.visibility = View.VISIBLE
+            rvRooms.visibility = View.GONE
+            btnSearchRooms.isSelected = false
+            btnSearchMessages.isSelected = true
+        }
+    }
+
     private fun setupSearchView() {
```
Comment (ivolarevic) [inline app/src/main/java/com/clover/studio/spikamessenger/ui/main/rooms/RoomsFragment.kt:75]: You can also add here with(binding) but this is optional

File: app/src/main/java/com/clover/studio/spikamessenger/ui/main/rooms/search/SearchAdapter.kt [hunk 12]
```diff
@@ -0,0 +1,85 @@
...
+                    }
+                }
+
+                binding.tvMessageDate.text = item.message.createdAt?.let {
+                    Tools.fullDateFormat(
+                        it
+                    )
+                }
+                binding.tvMessageContent.text = item.message.body?.text
+
```
Comment (ivolarevic) [inline app/src/main/java/com/clover/studio/spikamessenger/ui/main/rooms/search/SearchAdapter.kt:41]: Remove empty line

File: app/src/main/res/layout/fragment_chat.xml [hunk 13]
```diff
@@ -60,22 +60,61 @@
...
-    <androidx.constraintlayout.widget.ConstraintLayout
+    <LinearLayout
+        android:id="@+id/ll_search_rooms_messages"
+        android:layout_width="match_parent"
+        android:layout_height="wrap_content"
+        android:layout_margin="@dimen/sixteen_dp_margin"
+        android:background="@drawable/bg_message_received"
+        android:backgroundTint="@color/message_other"
+        android:orientation="horizontal"
+        android:padding="@dimen/five_dp_margin"
```
Comment (ivolarevic) [inline app/src/main/res/layout/fragment_chat.xml:71]: Try to set it to eight dp (if its not big difference)
(https://developer.android.com/develop/ui/views/layout/constraint-layout
All margins offered by the tool are factors of 8dp to help your views align to Material Design's 8dp square grid recommendations.)

File: app/src/main/res/layout/fragment_chat.xml [hunk 14]
```diff
@@ -60,22 +60,61 @@
...
+        android:visibility="gone"
+        app:layout_constraintEnd_toEndOf="parent"
+        app:layout_constraintStart_toStartOf="parent"
+        app:layout_constraintTop_toBottomOf="@id/sv_rooms_search">
+
+        <androidx.appcompat.widget.AppCompatButton
+            android:id="@+id/btn_search_rooms"
+            android:layout_width="0dp"
+            android:layout_height="@dimen/thirty_dp_margin"
```
Comment (ivolarevic) [inline app/src/main/res/layout/fragment_chat.xml:80]: Same as above, 32dp if its not big difference

File: app/src/main/res/layout/fragment_chat.xml [hunk 15]
```diff
@@ -60,22 +60,61 @@
...
+            android:background="@drawable/bg_button_bar_custom"
+            android:fontFamily="@font/montserrat"
+            android:text="@string/rooms"
+            android:textAllCaps="false"
+            android:textColor="@color/white" />
+
+        <androidx.appcompat.widget.AppCompatButton
+            android:id="@+id/btn_search_messages"
+            android:layout_width="0dp"
+            android:layout_height="@dimen/thirty_dp_margin"
```
Comment (ivolarevic) [inline app/src/main/res/layout/fragment_chat.xml:91]: Same as above

File: app/src/main/res/layout/item_message_search.xml [hunk 16]
```diff
@@ -0,0 +1,48 @@
+<?xml version="1.0" encoding="utf-8"?>
+<androidx.constraintlayout.widget.ConstraintLayout xmlns:android="http://schemas.android.com/apk/res/android"
+    xmlns:app="http://schemas.android.com/apk/res-auto"
+    android:layout_width="match_parent"
+    android:layout_height="wrap_content"
+    android:padding="@dimen/ten_dp_margin">
```
Comment (ivolarevic) [inline app/src/main/res/layout/item_message_search.xml:6]: If you can change it to eight dp. Also can you maybe use frame layout or linear or it is to complex?

File: app/src/main/res/values/colors.xml [hunk 17]
```diff
@@ -25,6 +25,7 @@
     <color name="error_background">#FFECEC</color>
     <color name="guideline">#E6E6E6</color>
     <color name="gray_transparent">#66F2F2F2</color>
+    <color name="gray">#F2F2F2</color>
```
Comment (ivolarevic) [inline app/src/main/res/values/colors.xml:28]: Check this color in dark and light mode. You also have this color: message_other with same code

File: app/src/main/res/layout/item_message_search.xml (hunk 16 above)
Comment (MatkoTom) [reply app/src/main/res/layout/item_message_search.xml:6]: We would have multiple nested linear layouts, I think it is better this way since there is only one constraint layout

File: app/src/main/res/values/colors.xml (hunk 17 above)
Comment (MatkoTom) [reply app/src/main/res/values/colors.xml:28]: Removed unused color, will add screenshot to PR

File: app/src/main/java/com/clover/studio/spikamessenger/ui/main/chat/ChatScreenActivity.kt (hunk 8 above)
Comment (MatkoTom) [reply app/src/main/java/com/clover/studio/spikamessenger/ui/main/chat/ChatScreenActivity.kt:87]: This was used for navigatin between chat screens when the in-app notification appears. The in-app notification has been removed from the app, so this is no longer used. I could remove it completely.

**Review by ivolarevic — APPROVED** [review]
LGTM

**Comment by ivolarevic** [comment]
Please change text color for rooms and messages. Use text_primary color.
